File: src/Moka.Red.Navigation/Sidebar/MokaSidebar.razor.cs
using System.Globalization;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Web;
using Moka.Red.Core.Utilities;

namespace Moka.Red.Navigation.Sidebar;

/// <summary>
///     A vertical sidebar navigation panel with collapsible, overlay, and mini modes.
///     Provides header, body (scrollable), and footer sections.
/// </summary>
public partial class MokaSidebar
{
	private const string ModulePath = "./_content/Moka.Red.Navigation/Sidebar/MokaSidebar.razor.js";

	private bool _disposed;
	private bool? _lastOpen;
	private ElementReference _navRef;
	private bool _open = true;

	// What the script was last told, so it hears each open and close of the overlay once.
	private bool _overlayShown;
	private bool _scrollLocked;

	/// <summary>Main body content (menu items, links, custom content).</summary>
	[Parameter]
	public RenderFragment? ChildContent { get; set; }

	/// <summary>Top area content (logo, app name).</summary>
	[Parameter]
	public RenderFragment? Header { get; set; }

	/// <summary>Bottom area content (user info, settings).</summary>
	[Parameter]
	public RenderFragment? Footer { get; set; }

	/// <summary>
	///     Whether the sidebar is open. Two-way bindable. Default true. An overlay sidebar closes itself
	///     on Escape and on a backdrop click, and reports it through <see cref="OpenChanged" />.
	/// </summary>
	[Parameter]
	public bool Open { get; set; } = true;

	/// <summary>Callback when <see cref="Open" /> changes.</summary>
	[Parameter]
	public EventCallback<bool> OpenChanged { get; set; }

	/// <summary>
	///     Mini-sidebar mode (icons only). Default false. The sidebar has no collapse control of its own,
	///     so the parent sets this.
	/// </summary>
	[Parameter]
	public bool Collapsed { get; set; }

	/// <summary>Full width when expanded. Default "240px".</summary>
	[Parameter]
	public string Width { get; set; } = "240px";

	/// <summary>Width when collapsed (mini mode). Default "56px".</summary>
	[Parameter]
	public string CollapsedWidth { get; set; } = "56px";

	/// <summary>
	///     When true, sidebar overlays content (mobile mode). Default false. While open it takes focus,
	///     keeps Tab inside it and closes on Escape, like a modal dialog.
	/// </summary>
	[Parameter]
	public bool Overlay { get; set; }

	/// <summary>Whether to show a right border. Default true.</summary>
	[Parameter]
	public bool Bordered { get; set; } = true;

	/// <summary>Whether to apply elevation shadow. Default false.</summary>
	[Parameter]
	public bool Elevated { get; set; }

	/// <summary>Position of the sidebar. Default Left.</summary>
	[Parameter]
	public MokaSidebarPosition Position { get; set; } = MokaSidebarPosition.Left;

	/// <inheritdoc />
	protected override string RootClass => "moka-sidebar";

	private string ResolvedWidth => Collapsed ? CollapsedWidth : Width;

	// An open overlay covers the page on a backdrop, so it behaves like a modal dialog.
	private bool OverlayOpen => Overlay && _open;

	// Closed, or collapsed to a rail with no width: the links are still in the page, but unseen.
	private bool OutOfView => !_open || (Collapsed && IsZeroLength(CollapsedWidth));

	private EventCallback<KeyboardEventArgs> KeyDownHandler => OverlayOpen
		? EventCallback.Factory.Create<KeyboardEventArgs>(this, HandleKeyDown)
		: default;

	/// <inheritdoc />
	protected override string CssClass => new CssBuilder(RootClass)
		.AddClass("moka-fill-height", !Overlay)
		.AddClass("moka-sidebar--open", _open)
		.AddClass("moka-sidebar--closed", !_open)
		.AddClass("moka-sidebar--collapsed", Collapsed)
		.AddClass("moka-sidebar--overlay", Overlay)
		.AddClass("moka-sidebar--bordered", Bordered)
		.AddClass("moka-sidebar--elevated", Elevated)
		.AddClass($"moka-sidebar--{MokaEnumHelpers.ToCssClass(Position)}")
		.AddClass(Class)
		.Build();

	// Closed, the sidebar drops its margin and padding. At width 0 the padding would still show as a
	// strip and the margin as a gap, and an overlay slid out by its own width would leave its margin
	// on screen.

	/// <inheritdoc />
	protected override string? CssStyle => new StyleBuilder()
		.AddStyle("margin", ResolvedMargin, _open)
		.AddStyle("padding", ResolvedPadding, _open)
		.AddStyle("border-radius", ResolvedRounding)
		.AddStyle("width", ResolvedWidth)
		.AddStyle(Style)
		.Build();

	/// <inheritdoc />
	protected override bool ShouldRender() => true;

	/// <inheritdoc />
	protected override void OnParametersSet()
	{
		base.OnParametersSet();

		// Open only seeds the state when the parent passes a new value. Copying it on every parent
		// render reopened an unbound overlay sidebar the user had just dismissed.
		if (_lastOpen != Open)
		{
			_lastOpen = Open;
			_open = Open;
		}
	}

	/// <inheritdoc />
	protected override async Task OnAfterRenderAsync(bool firstRender)
	{
		await base.OnAfterRenderAsync(firstRender);
		await SyncOverlayAsync();
		await SyncScrollLockAsync();
	}

	private async Task SyncOverlayAsync()
	{
		// Opening moves focus in and holds Tab, closing hands focus back. Both need the rendered
		// state: an open sidebar that is no longer inert, a closed one that is.
		bool shown = OverlayOpen;
		if (shown == _overlayShown)
		{
			return;
		}

		_overlayShown = shown;
		if (shown)
		{
			await SafeModuleInvokeVoidAsync(ModulePath, "openOverlay", _navRef);
		}
		else
		{
			// Overlay turned off while open, as in a layout that goes inline on wide screens: the
			// sidebar is still on screen, so focus stays where it is.
			await SafeModuleInvokeVoidAsync(ModulePath, "closeOverlay", _navRef, !_open);
		}
	}

	// The page behind an open overlay sidebar stays still, as behind a drawer. The lock is Core's
	// counted one, shared with dialogs and drawers, so the sidebar gives back only its own. The
	// flag flips before the call, so a render that finishes while a call is out cannot lock twice.
	// Kept apart from the focus calls: a sidebar removed while open is no longer in the page for
	// the script to find, but its lock still has to be given back.
	private async Task SyncScrollLockAsync()
	{
		bool wanted = OverlayOpen && !_disposed;
		if (wanted == _scrollLocked)
		{
			return;
		}

		_scrollLocked = wanted;
		await SafeModuleInvokeVoidAsync(ModulePath, wanted ? "lockBodyScroll" : "unlockBodyScroll");
	}

	/// <inheritdoc />
	protected override async ValueTask DisposeAsyncCore()
	{
		// Set first: a render still finishing must not take the lock again after it is given back.
		_disposed = true;
		await SyncScrollLockAsync();
		await base.DisposeAsyncCore();
	}

	private async Task HandleKeyDown(KeyboardEventArgs e)
	{
		if (e.Key == "Escape")
		{
			await CloseOverlayAsync();
		}
	}

	private Task HandleBackdropClick() => CloseOverlayAsync();

	private async Task CloseOverlayAsync()
	{
		if (!OverlayOpen)
		{
			return;
		}

		_open = false;
		await OpenChanged.InvokeAsync(false);
	}

	// "0", "0px", "0rem" and the like. Anything else, calc() included, counts as a width that shows.
	private static bool IsZeroLength(string? length)
	{
		ReadOnlySpan<char> value = length.AsSpan().Trim();
		int end = value.Length;
		while (end > 0 && (char.IsAsciiLetter(value[end - 1]) || value[end - 1] == '%'))
		{
			end--;
		}

		return end > 0
		       && double.TryParse(value[..end], NumberStyles.Float, CultureInfo.InvariantCulture, out double number)
		       && number == 0;
	}
}

File: src/Moka.Red.Navigation/Sidebar/MokaSidebar.razor.js
/**
 * MokaSidebar JS module - focus handling for an overlay sidebar.
 *
 * An open overlay sidebar covers the page on a backdrop, so it works like a modal dialog: opening
 * moves focus into it and keeps Tab inside it, and closing hands focus back to where it was.
 * Escape itself is handled in .NET. Dialogs get this from moka-dialog.js, but Navigation does not
 * reference Feedback, so the sidebar carries its own small version.
 *
 * The state lives in a WeakMap keyed by the sidebar, and the listener on the sidebar itself, so a
 * sidebar that leaves the page takes both with it and there is nothing to dispose.
 *
 * The page scroll lock is Core's shared, counted one, re-exported here so the sidebar's single
 * module serves every call. .NET pairs the calls itself, since a sidebar removed while open is
 * gone from the page by the time it is disposed.
 */

export { lockBodyScroll, unlockBodyScroll } from '../../Moka.Red.Core/moka-scroll-lock.js';

// Everything that can be in the tab order. A negative tabindex takes any of them out.
const FOCUSABLE = [
	'a[href]',
	'button:not([disabled])',
	'input:not([disabled]):not([type="hidden"])',
	'select:not([disabled])',
	'textarea:not([disabled])',
	'summary',
	'[contenteditable]:not([contenteditable="false"])',
	'[tabindex]'
].join(', ');

// Sidebar -> { onKeyDown, previous } while it is open as an overlay.
const openSidebars = new WeakMap();

function tabbableWithin(sidebar) {
	return Array.from(sidebar.querySelectorAll(FOCUSABLE))
		.filter(el => (!el.hasAttribute('tabindex') || el.tabIndex >= 0)
			&& !!(el.offsetWidth || el.offsetHeight || el.getClientRects().length));
}

// Tab and Shift+Tab go to the next or previous element in the tab order, counted in document order
// from the focused element, and wrap at the ends. Counting from focus keeps this right when focus
// sits on the sidebar itself or on an element taken out of the tab order.
function moveFocus(sidebar, backwards) {
	const tabbable = tabbableWithin(sidebar);
	if (tabbable.length === 0) {
		sidebar.focus();
		return;
	}

	const from = document.activeElement;
	const ordered = backwards ? tabbable.slice().reverse() : tabbable;
	const side = backwards ? Node.DOCUMENT_POSITION_PRECEDING : Node.DOCUMENT_POSITION_FOLLOWING;
	const next = from && sidebar.contains(from)
		? ordered.find(el => from.compareDocumentPosition(el) & side)
		: null;
	(next || ordered[0]).focus();
}

/**
 * Moves focus into an overlay sidebar that has just opened, and keeps Tab inside it until it
 * closes. Focus goes to an element marked autofocus or data-autofocus, else the first control,
 * else the sidebar itself. Content that already took focus inside the sidebar keeps it.
 * @param {HTMLElement} sidebar - The sidebar's nav element.
 */
export function openOverlay(sidebar) {
	if (!sidebar || openSidebars.has(sidebar)) return;

	const onKeyDown = e => {
		if (e.key !== 'Tab' || e.altKey || e.ctrlKey || e.metaKey || e.defaultPrevented) return;

		e.preventDefault();
		moveFocus(sidebar, e.shiftKey);
	};

	const active = document.activeElement;
	sidebar.addEventListener('keydown', onKeyDown);
	openSidebars.set(sidebar, { onKeyDown, previous: sidebar.contains(active) ? null : active });

	if (sidebar.contains(active)) return;

	const requested = sidebar.querySelector('[autofocus], [data-autofocus]');
	(requested || tabbableWithin(sidebar)[0] || sidebar).focus();
}

/**
 * Lets go of Tab and hands focus back to where it was before the sidebar opened. Focus only moves
 * when it is still in the sidebar, or nowhere because the closed sidebar went inert, so a close
 * that happens while the user is elsewhere does not pull focus away.
 * @param {HTMLElement} sidebar - The sidebar's nav element.
 * @param {boolean} [returnFocus=true] - False when the sidebar stays on screen, as an inline one.
 */
export function closeOverlay(sidebar, returnFocus = true) {
	const state = sidebar && openSidebars.get(sidebar);
	if (!state) return;

	sidebar.removeEventListener('keydown', state.onKeyDown);
	openSidebars.delete(sidebar);
	if (!returnFocus) return;

	const active = document.activeElement;
	const focusHere = !active || active === document.body || sidebar.contains(active);
	if (focusHere && state.previous && state.previous.isConnected) {
		state.previous.focus();
	}
}
